From dbbc9e087cfb24e99bfe3b692dc98803cd64edc3 Mon Sep 17 00:00:00 2001
From: Manoj Kumar Prabakaran <manojkumar.prabakaran@arm.com>
Date: Wed, 10 Apr 2024 13:26:46 +0530
Subject: [PATCH] neoverse-rd/rdv3 smcf_client: Add support for configuring MGI
 monitor mode

Added support for configuring the burst mode configuraton in
MGI monitor mode to enable the aux counter values in SMCF.

Change-Id: I8bd04937ed41ca53c91476477081f1aea481fa05
Signed-off-by: Manoj Kumar Prabakaran <manojkumar.prabakaran@arm.com>

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 .../module/smcf_client/src/mod_smcf_client.c    | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/product/neoverse-rd/rdv3/module/smcf_client/src/mod_smcf_client.c b/product/neoverse-rd/rdv3/module/smcf_client/src/mod_smcf_client.c
index aea215d530fd..de42ca562660 100644
--- a/product/neoverse-rd/rdv3/module/smcf_client/src/mod_smcf_client.c
+++ b/product/neoverse-rd/rdv3/module/smcf_client/src/mod_smcf_client.c
@@ -9,7 +9,6 @@
 
 #include <scp_cfgd_timer.h>
 #include <scp_cfgd_transport.h>
-
 #include <mod_amu_smcf_drv.h>
 #include <mod_platform_smcf.h>
 #include <mod_smcf.h>
@@ -32,6 +31,8 @@
 
 #define NUM_OF_CORE_COUNTERS 7
 #define MOD_NAME "[SMCF_CLIENT] "
+#define MODE_CONFIG_VALUE UINT32_C(0x51007000)
+#define MODE_REQ_INDEX UINT8_C(2)
 
 static unsigned int core_count;
 
@@ -64,7 +65,7 @@ static int smcf_client_init(
 static void alarm_callback(uintptr_t module_idx)
 {
     uint32_t mgi;
-    uint32_t num_of_mgis = 2 * core_count; // Each core has 2 MGI (AMU, DTS)
+    uint32_t num_of_mgis = core_count; // Each core has 2 MGI (AMU, DTS)
 
     for (mgi = 0; mgi < num_of_mgis; mgi++) {
         platform_smpling_api->start_data_sampling(
@@ -200,7 +201,8 @@ static int smcf_client_process_notification(
     }
 
     mgi_idx = fwk_id_get_element_idx(event->source_id);
-    FWK_LOG_INFO("CLIENT: Data successfully fetched for MGI[%u]", mgi_idx);
+
+    FWK_LOG_INFO(MOD_NAME "Data successfully fetched for MGI[%u]", mgi_idx);
 
     if (mgi_idx < core_count) { // AMUs
 
@@ -211,13 +213,13 @@ static int smcf_client_process_notification(
             NUM_OF_CORE_COUNTERS);
 
         if (status != FWK_SUCCESS) {
-            FWK_LOG_ERR(
+            FWK_LOG_ERR(MOD_NAME
                 "Failed to get amu data for core %u ERROR %d", mgi_idx, status);
             return status;
         }
         for (cntr_idx = 0; cntr_idx < NUM_OF_CORE_COUNTERS; ++cntr_idx)
-            FWK_LOG_ERR(
-                "CLIENT: MGI[%u] AMU_COUNTER[%u] data = %llu",
+            FWK_LOG_INFO(MOD_NAME
+                "MGI[%u] AMU_COUNTER[%u] data = %llu",
                 mgi_idx,
                 cntr_idx,
                 counter_value[cntr_idx]);
@@ -233,12 +235,13 @@ static int smcf_client_start(fwk_id_t id)
         return FWK_SUCCESS;
     }
 
-    num_of_mgis = 2 * core_count; // Each core has 2 MGI (AMU, DTS)
+    num_of_mgis = core_count;
     for (mgi = 0; mgi < num_of_mgis; mgi++) {
         fwk_notification_subscribe(
             mod_smcf_notification_id_new_data_sample_ready,
             FWK_ID_ELEMENT(FWK_MODULE_IDX_SMCF, mgi),
             id);
+	control_api->config_mode( (FWK_ID_SUB_ELEMENT(FWK_MODULE_IDX_SMCF, mgi, 0)),MODE_CONFIG_VALUE,MODE_REQ_INDEX);
     }
 
     return FWK_SUCCESS;
